logrotate.c

```diff
@@ -1344,49 +1342,114 @@ static long daysElapsed(const struct tm *now, const struct tm *last)
     return (long) (diff / (24 * 3600));
 }
 
+static int isSecureDirStep(const char *pathname, uid_t exempt_uid, gid_t exempt_gid)
+{
+    struct stat sb;
+
+    if (lstat(pathname, &sb) < 0)
+        return -errno;
+
+    if (S_ISLNK(sb.st_mode)) {
+        /* check whether symlink target is inside a secure directory */
+        char *canonpath;
+        int r;
+
+        canonpath = realpath(pathname, NULL);
+        if (!canonpath)
+            return -errno;
+
+        r = isSecureDir(canonpath, exempt_uid, exempt_gid);
+        free(canonpath);
+        return r;
+    }
+
+    if (!S_ISDIR(sb.st_mode))
+        return -ENOTDIR;
+
+    if ((sb.st_uid != ROOT_UID && sb.st_uid != exempt_uid) ||
+        ((sb.st_mode & S_IWGRP) && sb.st_gid != 0 && sb.st_gid != exempt_gid) ||
+        (sb.st_mode & S_IWOTH && !(sb.st_mode & S_ISVTX))) {
+        return 0;
+    }
+
+    return 1;
+}
+
+int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid)
+{
+    char *copy, *p;
+    int r;
+
+    if (pathname[0] != '/')
+        return -EINVAL;
+
+    copy = strdup(pathname);
+    if (!copy)
+        return -errno;
+
+    for (p = copy + 1; p; p = strchr(p + 1, '/')) {
+        char backup;
+
+        backup = *p;
+        *p = '\0';
+        r = isSecureDirStep(copy, exempt_uid, exempt_gid);
+        *p = backup;
+        if (r <= 0)
+            goto cleanup;
+    }
+
+    r = isSecureDirStep(pathname, exempt_uid, exempt_gid);
+
+cleanup:
+    free(copy);
+    return r;
+}
+
 static int findNeedRotating(const struct logInfo *log, unsigned logNum, int force)
 {
     struct stat sb;
-    struct logState *state = NULL;
+    struct logState *state;
     struct tm now;
 
     message(MESS_DEBUG, "considering log %s\n", log->files[logNum]);
 
     localtime_r(&nowSecs, &now);
 
-    /* Check if parent directory of this log has safe permissions */
-    if ((log->flags & LOG_FLAG_SU) == 0 && getuid() == ROOT_UID) {
-        char *ld;
-        char *logpath = strdup(log->files[logNum]);
-        if (logpath == NULL) {
+    /* Check if parent directories of this log has safe permissions */
+    if ((log->flags & LOG_FLAG_ALLOWINSECUREDIR) == 0 &&
+        (getuid() == ROOT_UID || getgid() == 0)) {
+        char *logpath;
+        const char *logdir;
+        uid_t exempt_uid = (log->flags & LOG_FLAG_SU) ? log->suUid : ROOT_UID;
+        gid_t exempt_gid = (log->flags & LOG_FLAG_SU) ? log->suGid : 0;
+        int r;
+
+        logpath = strdup(log->files[logNum]);
+        if (!logpath) {
             message_OOM();
             return 1;
         }
-        ld = dirname(logpath);
-        if (stat(ld, &sb)) {
+        logdir = dirname(logpath);
+        r = isSecureDir(logdir, exempt_uid, exempt_gid);
```

Comment:
True.
We can however not assume they are equal, e.g. `/var/log/app/*/*.log`.
One alternative would be to store the already checked directories in a hash table.

Reply:
Hash table is probably overkill and could have overhead in some workloads.  I was rather thinking about caching the result of the last checked directory only.  That is easy to implement, does not increase memory consumption over time, and it would cover the most common scenarios.